resource.go

```diff
@@ -60,13 +82,79 @@ func getResourceField(v interface{}) (f *structs.Field, err error) {
 	return
 }
 
+func getIdFromMap(m map[string]interface{}) (ret int, err error) {
+	if val, ok := m["id"].(int); !ok {
+		ret = 0
+	} else {
+		ret = val
+	}
+	return
+}
+
 // Retrieve Id field from interface.
 func resourceId(v interface{}) (ret int, err error) {
 	var f *structs.Field
+	invalidTypeErr := errors.New("Not a struct, map, or pointer to either")
+	ty := reflect.TypeOf(v)
 
-	if f, err = getResourceField(v); err == nil {
-		ret = f.Value().(int)
+	switch ty.Kind() {
+	case reflect.Map:
+		if m, ok := v.(map[string]interface{}); !ok {
+			err = errors.New("Invalid map")
+		} else {
+			return getIdFromMap(m)
+		}
+	case reflect.Struct:
+		if f, err = getResourceField(v); err == nil {
+			ret = f.Value().(int)
+		}
+	case reflect.Ptr:
+		return resourceId(reflect.Indirect(reflect.ValueOf(v)).Interface())
+	default:
+		err = invalidTypeErr
 	}
 
 	return
 }
+
+// Determine whether the struct's id will be omitted in json encoding.
```

Comment:
Why was this moved from `utils.go`? Seems like that's the better place for this function